lib/class-wp-theme-json-gutenberg.php

```diff
@@ -1354,32 +1362,127 @@ public function merge( $incoming ) {
 		$incoming_data    = $incoming->get_raw_data();
 		$this->theme_json = array_replace_recursive( $this->theme_json, $incoming_data );
 
-		// The array_replace_recursive algorithm merges at the leaf level.
-		// For leaf values that are arrays it will use the numeric indexes for replacement.
-		// In those cases, we want to replace the existing with the incoming value, if it exists.
+		/*
+		 * The array_replace_recursive algorithm merges at the leaf level,
+		 * but we don't want leaf arrays to be merged, so we overwrite it.
+		 *
+		 * For leaf values that are sequential arrays it will use the numeric indexes for replacement.
+		 * We rather replace the existing with the incoming value, if it exists.
+		 * This is the case of spacing.units.
+		 *
+		 * For leaf values that are associative arrays it will merge them as expected.
+		 * This is also not the behavior we want for the current associative arrays (presets).
+		 * We rather replace the existing with the incoming value, if it exists.
+		 * This happens, for example, when we merge data from theme.json upon existing
+		 * theme supports or when we merge anything coming from the same source twice.
+		 * This is the case of color.palette, color.gradients, color.duotone,
+		 * typography.fontSizes, or typography.fontFamilies.
+		 *
+		 */
 		$to_replace   = array();
 		$to_replace[] = array( 'spacing', 'units' );
-		foreach ( self::VALID_ORIGINS as $origin ) {
-			$to_replace[] = array( 'color', 'duotone', $origin );
-			$to_replace[] = array( 'color', 'palette', $origin );
-			$to_replace[] = array( 'color', 'gradients', $origin );
-			$to_replace[] = array( 'typography', 'fontSizes', $origin );
-			$to_replace[] = array( 'typography', 'fontFamilies', $origin );
+		foreach ( self::PRESETS_METADATA as $metadata ) {
+			foreach ( self::VALID_ORIGINS as $origin ) {
+				$to_replace[] = array_merge( $metadata['path'], array( $origin ) );
+			}
 		}
 
+		/**
+		 * We need to remove the incoming presets that have the same slug
+		 * than an existing default preset. This is the list of slugs present
+		 * in the global config we'll use to filter the incoming slugs.
+		 */
+		$slugs_global = self::get_slugs_not_to_override( $this->theme_json );
+
 		$nodes = self::get_setting_nodes( $this->theme_json );
 		foreach ( $nodes as $metadata ) {
 			foreach ( $to_replace as $property_path ) {
 				$path = array_merge( $metadata['path'], $property_path );
 				$node = _wp_array_get( $incoming_data, $path, null );
 				if ( isset( $node ) ) {
+					if ( 'theme' === $path[ count( $path ) - 1 ] ) {
```

Comment:
Could we check if $metadata['override'] is set on this if condition and avoid all the filter logic after (also removing the check for override inside get_slugs_not_to_override)?

Reply:
Pushed changes to do this.